smack-core/src/main/java/org/jivesoftware/smack/packet/Stanza.java

```diff
@@ -458,6 +468,144 @@ public ExtensionElement removeExtension(ExtensionElement extension)  {
         return removeExtension(extension.getElementName(), extension.getNamespace());
     }
 
+    /**
+     * Returns <tt>true</tt> if custom attributes management is enabled.
+     * @return <tt>true</tt> if custom attributes management is enabled.
+     */
+    public static boolean isCustomAttributesEnabled() {
+        return customAttributesEnabled;
+    }
+
+    /**
+     * Enables the possibility to add custom attributes to the stanza.
+     * <br/><b>Enabling this feature breaks the XMPP standard and may induce unexpected behaviors!
+     * <br/>If your use-case allow it, prefer using ExtensionElement instead, and if not, be sure to triple-check the result.</b>
+     */
+    public static void enableCustomAttributes() {
+        customAttributesEnabled = true;
+    }
+
+    private static void requireCustomAttributesEnabled(String methodName) {
+        if (!isCustomAttributesEnabled()) {
+            throw new IllegalStateException(
+                    "You cannot use " + methodName + " without enabling it first with enableCustomAttributes. " +
+                    "Ensure you are aware of the consequences of doing so.");
+        }
+    }
+
+    /**
+     * Replaces all custom attributes of the stanza.
+     * @param attributes a map (name/value) of all the custom attributes of this stanza. SHould not be {@null}
+     * @throws IllegalStateException if custom attributes management is not enabled.
+     */
+    public void setCustomAttributes(Map<String, String> attributes) {
+        requireCustomAttributesEnabled("setCustomAttributes");
+        assert attributes != null;
+
+        synchronized (customAttributes) {
+            customAttributes.clear();
+            customAttributes.putAll(attributes);
+        }
+    }
+
+    /**
+     * Adds a custom attribute to the stanza. If a custom attribute with the same name is already present, its value is updated.
+     * @param attributeName the custom attribute name.
+     * @param attributeValue the custom attribute value.
+     * @throws IllegalStateException if custom attributes management is not enabled.
+     */
+    public void setCustomAttribute(String attributeName, String attributeValue) {
+        requireCustomAttributesEnabled("setCustomAttribute");
```

Comment:
You need to check in the set* methods that the custom attribute name doesn't clash with a common attribute name to prevent users from producing invalid XML. Throw an IllegalArgumentException if so.